components/scream/src/physics/share/physics_universal_impl.hpp

```diff
@@ -0,0 +1,118 @@
+#ifndef PHYSICS_UNIVERSAL_IMPL_HPP
+#define PHYSICS_UNIVERSAL_IMPL_HPP
+
+#include "physics_functions.hpp" // for ETI only but harmless for GPU
+#include "physics/share/physics_constants.hpp"
+
+namespace scream {
+namespace physics {
+
+/*
+ * Implementation of universal physics functions. Clients should NOT #include
+ * this file, #include physics_functions.hpp instead.
+ */
+
+//-----------------------------------------------------------------------------------------------//
+// Applies Exners Function which follows:
+//   Exner = (P/P0)^(Rd/Cp),
+// where,
+//   P  is the pressure at this location, Pa
+//   P0 is a reference pressure, Pa
+//   Rd is the gas constant, J/K
+//   Cp is heat capacity of dry air, J/Ki
+// All universal constants, P0, Rd, and Cp, are defined in physics_constants.hpp
+// Note: Another experssion for Exner is,
+//   Exner = T/th
+// whre,
+//   T  is the temperature, K
+//   th is the potential temperature, K
+template <typename S, typename D>
+KOKKOS_FUNCTION
+typename Functions<S,D>::Spack
+Functions<S,D>::get_exner(const Spack& P, const Smask& range_mask)
+{
+  Spack result;
+  
+  static constexpr Scalar p0     = C::P0;
+  static constexpr Scalar rd     = C::RD;
+  static constexpr Scalar inv_cp = C::INV_CP;
+  const Spack exner = pow( P/p0, rd*inv_cp );
+  // Check that there are no obvious errors in the result.
+  auto is_nan_exner = isnan(exner) && range_mask;
+  EKAT_KERNEL_REQUIRE_MSG(!(is_nan_exner.any()), "Error in get_exner, Exner has NaN values.\n"); // exit with an error message
```

Comment:
REQUIRE implies this is always on, even in non-debug builds. Is that intended? For errors that are to be caught only in debug builds, use EKAT_KERNEL_ASSERT_MSG. (There tends to be a convention that ASSERT implies dependence on NDEBUG whereas REQUIRE or similarly named macros are always enabled.)

Reply:
TBH, I'm not sure.  My impulse is that we should never let a NaN exner value get through, debug or not debug build.  But if there is a good performance reason to only catch in DEBUG builds I'm happy to switch to ASSERT.  Presumably, anyone who is getting errors in a non-debug build should be checking with a debug build eventually.

Reply:
I agree with Andrew. I think kernels should only contain debug checks. Otherwise, where do we draw the line? If we check that one thing is not NaN, why not checking all the inputs every time? The cost of all those checks on all entries of the arrays can quickly pile up.(1)

Release builds should contain checks on scalar quantities (meaning quantities that are not arrays over the pb geo/time dimensions). E.g., if you compute some reduction (max/min) of a distributed quantity as part of the algo, it is ok to check the reduction value.

1. Note: I'm ok if you check all the inputs _in a debug build_. Debug runs can contain as many checks as we want.

Reply:
ok cool. I'm happy to change this to ASSERT. I'll do that shortly and push a new version.

Reply:
My apologies for going with REQUIRE, I noticed just now that I was using the `check_temperature` routine in `physics_saturation_impl.hpp` as an example.   But since that is a case where you would want an error if the temperature had incorrect values it makes sense that it would throw an error in debug and non-debug builds alike.

Reply:
Note that if you switch, you should make the error checking all inside the ASSERT call, like this:
```c++
EKAT_KERNEL_ASSERT_MSG(!(isnan(exner) && range_mask).any(), "msg");
```
That way, when NDEBUG is defined, the whole error-checking calculation is removed.